File: JavaScript/0152. Maximum Product Subarray.js
// Given an integer array nums, find the contiguous subarray within an array (containing at least one number) which has the largest product.
//
// Example 1:
//
// Input: [2,3,-2,4]
// Output: 6
// Explanation: [2,3] has the largest product 6.
//
// Example 2:
//
// Input: [-2,0,-1]
// Output: 0
// Explanation: The result cannot be 2, because [-2,-1] is not a subarray.

/**
 * @param {number[]} nums
 * @return {number}
 */

// 1) Brute force
// Time O(n^2)
// Space O(1)
const maxProduct1 = (nums) => {
  if (nums.length === 1) return nums[0];

  let max = -Infinity;
  for (let i = 0; i < nums.length; i++) {
    let p = nums[i];
    max = Math.max(max, p);

    for (let j = i + 1; j < nums.length; j++) {
      p *= nums[j];
      max = Math.max(max, p);
    }
  }
  return max;
};

// 2)
// Similar
// 53. Maximum Subarray
//
// Time O(n)
// Space O(1)
const maxProduct = (nums) => {
  let max = -Infinity;
  let currMin = 1;
  let currMax = 1;
  for (const n of nums) {
    [currMin, currMax] = [
      Math.min(n, currMin * n, currMax * n),
      Math.max(n, currMin * n, currMax * n),
    ];
    max = Math.max(max, currMax);
  }
  return max;
};

File: Python/0152. Maximum Product Subarray.py
# Given an integer array nums, find a contiguous non-empty subarray within the array that has the largest product, and return the product.
# The test cases are generated so that the answer will fit in a 32-bit integer.
# A subarray is a contiguous subsequence of the array.
#
# Example 1:
#
# Input: nums = [2,3,-2,4]
# Output: 6
# Explanation: [2,3] has the largest product 6.
#
# Example 2:
#
# Input: nums = [-2,0,-1]
# Output: 0
# Explanation: The result cannot be 2, because [-2,-1] is not a subarray.
#
# Constraints:
#
# 1 <= nums.length <= 2 * 10^4
# -10 <= nums[i] <= 10
# The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.


# Dynamic Programming
# Time O(n)
# Space O(1)
#
# Similar
# 53. Maximum Subarray
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]

        cur_max = nums[0]
        cur_min = nums[0]
        m = nums[0]

        for i in range(1, len(nums)):
            if nums[i] < 0:
                cur_max, cur_min = cur_min, cur_max
            cur_max = max(nums[i], cur_max * nums[i])
            cur_min = min(nums[i], cur_min * nums[i])
            m = max(m, cur_max)
        return m
